Providers/RustSyncManager.swift

```diff
@@ -0,0 +1,553 @@
+                                kind: self.toSyncManagerDeviceType(
+                                    deviceType: device.deviceType)))
+
+                        self.beginSyncing()
+                        self.syncManagerAPI.sync(params: params) { syncResult in
+                            guard syncResult.status == ServiceStatus.ok else {
+                                deferred.fill(Maybe(success: syncResult))
+                                self.endSyncing(syncResult)
+                                return
+                            }
+
+                            // Save the persisted state
+                            self.prefs
+                                .setString(syncResult.persistedState,
+                                           forKey: PrefsKeys.RustSyncManagerPersistedState)
+
+                            let declinedEngines = String(describing: syncResult.declined ?? [])
+                            let telemetryData = syncResult.telemetryJson ??
+                                "(No telemetry data was returned)"
+                            let telemetryMessage = "\(String(describing: telemetryData))"
+                            let syncDetails = ["status": "\(syncResult.status)",
+                                               "declinedEngines": "\(declinedEngines)",
+                                               "telemetry": telemetryMessage]
+
+                            self.logger.log("Finished syncing",
+                                            level: .info,
+                                            category: .sync,
+                                            extra: syncDetails)
+
+                            // Save declined/enabled engines - we assume the engines
+                            // not included in the returned `declined` property of the
+                            // result of the sync manager `sync` are enabled.
+                            let updateEnginePref:
+                            (String, Bool) -> Void = { engine, enabled in
+                                let enabledPref = "engine.\(engine).enabled"
+                                self.prefsForSync.setBool(enabled, forKey: enabledPref)
+
+                                let stateChangedPref = "engine.\(engine).enabledStateChanged"
+                                self.prefsForSync.setObject(nil, forKey: stateChangedPref)
+                            }
+
+                            if let declined = syncResult.declined {
+                                RustTogglableEngines.forEach({
+                                    if declined.contains($0) {
+                                        updateEnginePref($0, false)
+                                    } else {
+                                        updateEnginePref($0, true)
+                                    }
+                                })
+                            } else {
+                                RustTogglableEngines.forEach({
```

Comment:
This branch should be removed - it is nullable, and `null` will just mean we failed before we could determine what the error is. The branch above (ie, where you iterate over `declined` if it is non-null) should be done even if the success value isn't ok.